### PR TITLE
Introduce QuarkusProdModeTest

### PR DESCRIPTION
**Pull request by geoand**
The main motivation is to get rid of maven invoker style tests that are pretty much impossible to debug

### CODE REVIEW

**Comment by geoand** [comment]
I would appreciate if @stuartwdouglas and @aloubyansky would review the `QuarkusProdModeTest` part and if @iocanel could vet for the kubernetes test migration :)

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 1]
```diff
@@ -415,43 +434,71 @@ private void doThinJarGeneration(CurateOutcomeBuildItem curateOutcomeBuildItem,
 
         final List<AppDependency> appDeps = curateOutcomeBuildItem.getEffectiveModel().getUserDependencies();
 
-        copyLibraryJars(transformedClasses, libDir, classPath, appDeps);
+        Map<AppDependency, List<Path>> classesOfNonJarDependencies = copyLibraryJars(transformedClasses, libDir, classPath,
+                appDeps, services);
```
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java]: If we are doing that, why not copy them while walking the dependency tree?

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java]: Very good point! I'll update the PR.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java]: Fixed

**Review by aloubyansky — APPROVED** [review]
Looks good to me.

**Comment by geoand** [comment]
Darn, it seems like something is wrong with Java 11+. I'll have to take a look

**Review by stuartwdouglas — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 2]
```diff
@@ -284,6 +237,72 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
         return new JarBuildItem(runnerJar, originalJar, null);
     }
 
+    private boolean isAppDepAJar(AppArtifact artifact) {
+        return "jar".equals(artifact.getType());
+    }
+
+    private boolean isAppDepPathAJar(AppArtifact artifact) {
+        return artifact.getPath().getFileName().toString().endsWith(".jar");
```
[the remark this answers is not in the document]
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java]: Makes perfect sense, fixed

**Review by iocanel — APPROVED** [review]
Personally, I prefer using groovy scripts for verifying the structure of the generated documents, as its less verbose than java. Of course, this is purely subjective and most probably the less popular opinion.

Still, I like that we provide this option to our users and since code-wise seems solid, here's my +1.

**Comment by geoand** [comment]
I agree that groovy scripts are more readable and certainly easier to write, but my major issue with them is that I can't debug them... 
And the whole Maven Invoker dance adds an extra layer of complication for anyone trying to debug tests

**Comment by iocanel** [comment]
> I agree that groovy scripts are more readable and certainly easier to write, but my major issue with them is that I can't debug them...
> And the whole Maven Invoker dance adds an extra layer of complication for anyone trying to debug tests

True!

**Pull request merged**